File: datasets/all/AI/binary_search_1.py
def binary_search(arr, target):
    low = 0
    high = len(arr) - 1

    while low <= high:
        mid = (low + high) // 2
        guess = arr[mid]

        if guess == target:
            return mid  # 要素が見つかった場合、その位置（インデックス）を返す
        if guess > target:
            high = mid - 1  # 探索範囲を下半分に絞る
        else:
            low = mid + 1  # 探索範囲を上半分に絞る

    return None  # 要素が見つからなかった場合

# テストコード
if __name__ == '__main__':
    my_list = [1, 3, 5, 7, 9, 11, 13, 15, 17]
    print(binary_search(my_list, 3))  # 出力: 1
    print(binary_search(my_list, -1)) # 出力: None

File: AICodeDetector/train2/AI/binarySearch1.go
package main

import (
	"fmt"
)

// binarySearchRecursive performs binary search on a sorted array recursively.
// It returns the index of the target if found, else -1.
func binarySearchRecursive(arr []int, target, low, high int) int {
	if high < low {
		return -1
	}

	mid := low + (high-low)/2
	if arr[mid] == target {
		return mid
	} else if arr[mid] < target {
		return binarySearchRecursive(arr, target, mid+1, high)
	} else {
		return binarySearchRecursive(arr, target, low, mid-1)
	}
}

func main() {
	arr := []int{2, 5, 8, 12, 16, 23, 38, 56, 72, 91}
	target := 56

	result := binarySearchRecursive(arr, target, 0, len(arr)-1)
	if result != -1 {
		fmt.Printf("Element found at index: %d\n", result)
	} else {
		fmt.Println("Element not found in the array.")
	}
}
